src/coreclr/jit/importer.cpp

```diff
@@ -4007,6 +4013,29 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsKnownConstant:
+            {
+                if (opts.OptimizationEnabled())
+                {
+                    GenTree* op1 = impPopStack().val;
+                    if (op1->OperIsConst())
```

Comment:
String literals get fetched via extra indirections. How does this work? Where do the extra indirections get the `GTK_CONST` flag?

Reply:
Ah ok, the PR description mentions that this does not work.

Reply:
`string.Empty` and `static readonly string` don't work (but can be arranged). Normal string literals do work because we import them as GT_CNS_STR and expand to indirections later in morph

Reply:
Sadly for diffs, we prefer `string.Empty` over `""` in BCL, also all comparisons against empty strings were manually converted to `str?.Length == 0` over the codebase in https://github.com/dotnet/runtime/pull/36443/files

Reply:
@stephentoub ah makes sense for me now why your analyzer found 100 cases in https://github.com/dotnet/runtime/issues/63719

Reply:
Should `CA1820` be removed when all these changes defeat its perf improvement reasoning?

Reply:
> Should CA1820 be removed when all these changes defeat its perf improvement reasoning?

It's already disabled by default.  If there ends up being zero benefit after these improvements (including subsequent ones to get string.Empty and the like), we could consider deleting it entirely from the SDK.  I wouldn't rush to that, though.